VirtoCommerce.Storefront/Domain/StaticContent/StaticContentItemBuilder.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using AutoRest.Core.Utilities;
+using Markdig;
+using VirtoCommerce.Storefront.Model.StaticContent;
+
+namespace VirtoCommerce.Storefront.Domain
+{
+    public partial class StaticContentItemBuilder : IStaticContentItemBuilder
+    {
+        internal static readonly Regex headerRegExp = new Regex(@"(?s:^---(.*?)---)");
+        internal static readonly string[] extensions = new[] { ".md", ".liquid", ".html" };
+
+        private readonly IStaticContentItemFactory _factory;
+
+        private readonly List<IContentItemVisitor> _prepareVisitors = new List<IContentItemVisitor>
+        {
+            new LangVisitor(),
+            new YamlMetadataVisitor(),
+            new PageMetadataVisitor(),
+        };
+
+        private readonly List<IContentItemVisitor> _metadataVisitors = new List<IContentItemVisitor>
+        {
+            new BlogExcerptMetadataVisitor(),
+            new BlogMetadataVisitor(),
+            new ContentPageMetadataVisitor(),
+            new MetadataVisitor()
+        };
+
+        private readonly List<IContentItemVisitor> _contentVisitors = new List<IContentItemVisitor>
+        {
+            new MarkdownVisitor(new MarkdownPipelineBuilder().UseAdvancedExtensions().Build()),
+            new PageContentVisitor(),
+        };
+
+        private readonly List<IContentItemVisitor> _postVisitors = new List<IContentItemVisitor> { new UrlsVisitor() };
+
+
+        public StaticContentItemBuilder(IStaticContentItemFactory factory)
+        {
+            _factory = factory;
+        }
+
+        public ContentItem BuildFrom(string baseStoreContentPath, string blobRelativePath, string content)
+        {
+            var contentItem = _factory.GetItemFromPath(blobRelativePath);
+            var visitedContent = content;
+            Action<List<IContentItemVisitor>> runParsers = parsers =>
+            {
+                parsers.Where(x => x.Suit(contentItem))
+                    .ForEach(x => visitedContent = x.ReadContent(blobRelativePath, visitedContent, contentItem));
+            };
+            if (contentItem != null)
+            {
+                contentItem.StoragePath = "/" +
+                    (string.IsNullOrEmpty(baseStoreContentPath) ?
+                        blobRelativePath :
+                        blobRelativePath.Replace(baseStoreContentPath + "/", String.Empty)
+                    ).TrimStart('/');
+                contentItem.FileName = Path.GetFileName(blobRelativePath);
+                new List<List<IContentItemVisitor>>
```

Comment:
why do we need to run all visitors again? Is it only to take into account the storage patch changes that are introduced in line 59? I think it is can be optimized to do only by single bypass

Reply:
it is a single now. on line 52 it is a functional variable to run visitors